library/alloc/src/collections/binary_heap.rs

```diff
@@ -1568,20 +1568,36 @@ impl<T: Ord, I: IntoIterator<Item = T>> SpecExtend<I> for BinaryHeap<T> {
     }
 }
 
+impl<T: Ord> SpecExtend<Vec<T>> for BinaryHeap<T> {
+    fn spec_extend(&mut self, ref mut other: Vec<T>) {
+        let start = self.data.len();
+        self.data.append(other);
+        self.rebuild_tail(start);
+    }
+}
+
 impl<T: Ord> SpecExtend<BinaryHeap<T>> for BinaryHeap<T> {
     fn spec_extend(&mut self, ref mut other: BinaryHeap<T>) {
         self.append(other);
     }
 }
 
+// RebuildTail is a struct that ensure `rebuild_tail` in `fn extend_desugared`
+// is executed
+struct RebuildTail<'a, T: Ord>(&'a mut BinaryHeap<T>, usize);
+impl<'a, T: Ord> Drop for RebuildTail<'a, T> {
+    #[inline]
+    fn drop(&mut self) {
+        self.0.rebuild_tail(self.1);
+    }
+}
+
 impl<T: Ord> BinaryHeap<T> {
     fn extend_desugared<I: IntoIterator<Item = T>>(&mut self, iter: I) {
-        let iterator = iter.into_iter();
-        let (lower, _) = iterator.size_hint();
-
-        self.reserve(lower);
-
-        iterator.for_each(move |elem| self.push(elem));
+        let start = self.data.len();
+        let extend = RebuildTail(self, start);
+        extend.0.data.extend(iter);
+        // `self.rebuild_tail(start)` is done by `drop(extend)`
```

Comment:
I'm a bit worried about the panic behaviour. I can see that you want to maintain the heap property even when the iterator panics, like the old implementation did, however contrary to the old implementation this approach could result in an abort if both the iterator and the comparison function of `T` panic.

Reply:
comparison function of T panic could be fine since we could not guarantee the heap order without comparison function.

Since we have the bound <T:Ord>
(Ord: Trait for types that form a total order. exactly one of a < b, a == b or a > b is true)
which means, comparison function of T should not panic.

If it panic, T could not satisfy the Ord trait, thus we could not guarantee a correct result.

Reply:
> comparison function of T panic could be fine since we could not guarantee the heap order without comparison function.

My problem is not with the incorrect order, rather that if both the iterator and the comparison function panics then the process aborts, without giving the user the ability to catch it. 

Hence the question: do we value more the heap property in case of a single panic or avoiding an abort in case of a double panic?
In my opinion not preserving the heap ordering in case of panic is not a big deal because any other method on `BinaryHeap` does that, including your method. Avoiding an abort instead might be more important because there are legitimate cases where you don't want to abort if one task panics, for example web frameworks.

Reply:
This has been marked as resolved, what's the conclusion of the discussion?

Reply:
> use `AssertUnwindSafe` or implement `UnwindSafe`, which are both kind of asserting that you don't care about potential inconsistencies.

As [the book](https://doc.rust-lang.org/book/ch09-03-to-panic-or-not-to-panic.html) said:
> If someone calls your code and passes in values that don’t make sense, the best choice might be to call panic! and alert the person using your library to the bug in their code so they can fix it during development. 

What's more, keep potential logic error in the program is not a good choice, and a simply change could fix that issue:
```
impl Ord for NotNaN {
    fn cmp(&self, other: &Self) -> Ordering {
        self.0.partial_cmp(&other.0).unwrap_or(Ordering::Equal)
    }
}
```
if one does not care about that NaN would lead to an invaild state, that modification is totally OK.
Otherwise, panic could be a better choice that it would notify developer that they made a mistake

Reply:
> As the book said:
> > If someone calls your code and passes in values that don’t make sense, the best choice might be to call panic! and alert the person using your library to the bug in their code so they can fix it during development.

That's completly unrelated to `AssertUnwindSafe` though. I'm not saying you shouldn't panic, do it if you think it's the right choice, but that alone won't make you observe broken logic invariants of e.g. `BinaryHeap`. To observe them you need to use `AssertUnwindSafe` or implement `UnwindSafe`, and by doing that you're asserting that you don't care if some logic invariants are broken.

> What's more, keep potential logic error in the program is not a good choice, and a simply change could fix that issue:
> ```
> impl Ord for NotNaN {
>     fn cmp(&self, other: &Self) -> Ordering {
>         self.0.partial_cmp(&other.0).unwrap_or(Ordering::Equal)
>     }
> }
> ```

That's the opposite of "not keeping a logic error". That way if a logic error happens it will silently succeed, even though it shouldn't! A panic there is the right choice because if a `NaN` reached that point then the program reached a state that shouldn't have been possible and some other piece of code had a bug.